File: paicord-rs/PaicordLibBridge/Sources/DiscordModels/Protobuf/PreloadedUserSettings.swift
import PaicordLib

class BridgedGuildFolderVec {
    var inner: [DiscordProtos_DiscordUsers_V1_PreloadedUserSettings.GuildFolder]

    init(inner: [DiscordProtos_DiscordUsers_V1_PreloadedUserSettings.GuildFolder]) {
        self.inner = inner
    }

    func len() -> UInt {
        return UInt(self.inner.count)
    }

    func get_unchecked(index: UInt) -> GuildFolderRust {
        return GuildFolderRust(folder: self.inner[Int(index)])
    }
}

extension GuildFolderRust {
    init(folder: DiscordProtos_DiscordUsers_V1_PreloadedUserSettings.GuildFolder) {
        var guildIds: RustVec<UInt64> = RustVec()

        for guildId in folder.guildIds {
            guildIds.push(value: guildId)
        }

        self.guild_ids = guildIds
        self.id = if folder.hasID {
            folder.id.value
        } else {
            nil
        }

        self.name = if folder.hasName {
            folder.name.value.intoRustString()
        } else {
            nil
        }

        self.color = if folder.hasColor {
            DiscordColorRust(inner: Int32(folder.color.value))
        } else {
            nil
        }
    }
}

class BridgedGuildFolders {
    var inner: DiscordProtos_DiscordUsers_V1_PreloadedUserSettings.GuildFolders

    init(inner: DiscordProtos_DiscordUsers_V1_PreloadedUserSettings.GuildFolders) {
        self.inner = inner
    }

    func get_folders() -> BridgedGuildFolderVec {
        return BridgedGuildFolderVec(inner: self.inner.folders)
    }

    func get_guild_positions() -> RustVec<UInt64> {
        self.inner.guildPositions.withUnsafeBufferPointer({ buf in
            return  make_rust_vec_u64(buf)
        })
    }
}

class BridgedPreloadedUserSettings {
    var inner: DiscordProtos_DiscordUsers_V1_PreloadedUserSettings

    init(inner: DiscordProtos_DiscordUsers_V1_PreloadedUserSettings) {
        self.inner = inner
    }

    func get_guild_folders() -> BridgedGuildFolders {
        return BridgedGuildFolders(inner: self.inner.guildFolders)
    }
}

File: paicord-rs/src/discord_models/protobuf/preloaded_user_settings.rs
use std::sync::Arc;

use crate::ffi::{BridgedGuildFolders, BridgedPreloadedUserSettings, GuildFolderRust};

pub type GuildFolder = GuildFolderRust;

#[derive(Clone, Debug)]
pub struct PreloadedUserSettings {
    pub guild_folders: GuildFolders,
}

impl PreloadedUserSettings {
    pub(crate) fn new(inner: BridgedPreloadedUserSettings) -> Self {
        Self {
            guild_folders: GuildFolders::new(inner.get_guild_folders()),
        }
    }
}

#[derive(Clone, Debug)]
pub struct GuildFolders {
    pub folders: Vec<GuildFolder>,
    pub guild_positions: Vec<u64>,
}

impl GuildFolders {
    pub(crate) fn new(inner: BridgedGuildFolders) -> Self {
        let folders = inner.get_folders();
        let folders = (0..folders.len())
            .map(|i| folders.get_unchecked(i))
            .collect();
        let guild_positions = inner.get_guild_positions();

        Self {
            folders,
            guild_positions,
        }
    }
}

impl std::fmt::Debug for BridgedPreloadedUserSettings {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("BridgedPreloadedUserSettings").finish()
    }
}

impl std::fmt::Debug for GuildFolder {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("GuildFolderRust")
            .field("guild_ids", &self.guild_ids)
            .field("id", &self.id)
            .field("name", &self.name)
            .field("color", &self.color)
            .finish()
    }
}

unsafe impl Send for BridgedPreloadedUserSettings {}
unsafe impl Sync for BridgedPreloadedUserSettings {}
